keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,171 @@
+			Name:               m.qosContainersInfo.Burstable,
+			ResourceParameters: &ResourceConfig{},
+		},
+		v1.PodQOSBestEffort: {
+			Name:               m.qosContainersInfo.BestEffort,
+			ResourceParameters: &ResourceConfig{},
+		},
+	}
+    ...
+	// ADDED
+	var configs []*CgroupConfig
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.MemoryQOS && libcontainercgroups.IsCgroup2UnifiedMode() {
+		configs = m.setMemoryQoSConfig()
+	}
+	...
+	configs = append(configs, qosConfigs[v1.PodQOSGuaranteed], qosConfigs[v1.PodQOSBurstable], qosConfigs[v1.PodQOSBestEffort])
+
+	for _, config := range configs {
+		err := m.cgroupManager.Update(config)
+		if err != nil {
+			klog.Errorf("[ContainerManager]: Failed to update QoS cgroup configuration")
+			return err
+		}
+	}
+
+	klog.V(4).Infof("[ContainerManager]: Updated QoS cgroup configuration")
+	return nil
+}
+```
+
+Add a new function `setMemoryQoSConfig` to set memory qos in cgroup config when pod is `Guaranteed` or `Burstable`.
+```
+pkg/kubelet/cm/qos_container_manager_linux.go
+// ADDED
+func (m *qosContainerManagerImpl) setMemoryQoSConfig() []*CgroupConfig {
+	pods := m.activePods()
+	cgs := make([]*CgroupConfig, 0)
+	for i := range pods {
+		pod := pods[i]
+		qosClass := v1qos.GetPodQOS(pod)
+		// if besteffort, we do nothing
+		if qosClass == v1.PodQOSBestEffort {
+			continue
+		}
+
+		cgroupConfig := CgroupConfig{}
+		cgroupConfig.Name = GetPodContainerName(pod)
+		if qosClass == v1.PodQOSGuaranteed {
+			cgroupConfig.ResourceParameters.Unified[MemoryMin] = reqs.Memory()
+		} else if qosClass == v1.PodQOSBurstable {
+			cgroupConfig.ResourceParameters.Unified[MemoryLow] = reqs.Memory()
+		}
+		cgs = append(cgs, &cgroupConfig)
+	}
+
+	return cgs
+}
+```
+
+### CGroups v2 Support
+After Kubernetes v1.19, kubelet can indentify cgroups v2 and do the convention. Since [v1.0.0-rc93](https://github.com/opencontainers/runc/releases/tag/v1.0.0-rc93), runc supports `Unified` to pass through cgroups v2 parameters. So we use this variable to pass `memory.low / memory.min` when cgroups v2 mode is detected.
+
+```
+pkg/kubelet/cm/cgroup_manager_linux.go
+func (m *cgroupManagerImpl) toResources(resourceConfig *ResourceConfig) *libcontainerconfigs.Resources {
+    ...
+    if libcontainercgroups.IsCgroup2UnifiedMode() {
+		if v, ok := resources.Unified[MemoryLow]; ok {
+			resources.Unified["memory.low"] = v
+		}
+		if v, ok := resources.MemoryExtras[MemoryMin]; ok {
+			resources.Unified["memory.min"] = v
+		}
+	}
+    ...
+}
+```
+
+cgroup hierarchy would be:
+
+```
+// guaranteed pods
+/cgroup2/kubepods/pod<UID>/memory.min = sum(pod.spec.containers.resources.requests[memory])
+/cgroup2/kubepods/pod<UID>/memory.max = sum(pod.spec.containers.resources.limits[memory])
+
+// burstable pods
+/cgroup2/kubepods/burstable/pod<UID>/memory.low = sum(pod.spec.containers.resources.requests[memory])
+/cgroup2/kubepods/burstable/pod<UID>/memory.max = sum(pod.spec.containers.resources.limit[memory])
+```
```

Comment:
We should add a section on testing this.

Reply:
> We should add a section on testing this.

ok, I will update.

Reply:
Done